relso/nn/transformers.py

```diff
@@ -82,10 +85,135 @@ def forward(self, x):
         return x
 
 
+
+
+
+
+####################################################################
+#### SUPER-DIRTY CODE FOR LINFORMER MULTI-HEAD ATTENTION ##########
+
+class MultiHeadLinformerAttention(nn.Module):
+    def __init__(self, input_dim, input_seq_len, h_times_embed_dim, num_heads, k):
+        """
+            input_dim - embedding vect dim * num_heads
+            input_seq_len - length of input sequences (must be constant among sequences)
+            h_times_embed_dim - d_v * num_heads (d_v = d_k = d_q; dim of single query/key/val vector)
+            num_heads - nuber of heads in attention block
+            k - length of sequence after E/F projections
+        """
+        super().__init__()
+        assert h_times_embed_dim % num_heads == 0, "Embedding dimension must be 0 modulo number of heads." ## as in original relso implementation
+
+        self.num_heads = num_heads
+        self.d_v = h_times_embed_dim // num_heads #(d_v = d_k = d_q)
+
+
+        # Projecting input to Q,K,V matrices for all heads (stacked for efficiency)
+        self.qkv_proj = nn.Linear(input_dim, h_times_embed_dim*3, bias=False)
+
+        #Projecting sequence of keys -- from length=input_seq_len to length=k
+        self.k_projections = nn.ModuleList([nn.Linear(input_seq_len, k, bias=False) for i in range(num_heads)]) # projection matrices Ei, i=1,..,num_heads
+        
+        #Projecting sequence of values -- from length=input_seq_len to length=k
+        self.v_projections = nn.ModuleList([nn.Linear(input_seq_len, k, bias=False) for i in range(num_heads)]) # projection matrices Fi, i=1,..,num_heads
+
+        ##### pasted from MultiHeadAttention ########
+        self.o_proj = nn.Linear(embed_dim, embed_dim)
+        #self._reset_parameters()
+        ############################################
+
+
+
+    # def _reset_parameters(self):
+    #     # Original Transformer initialization, see PyTorch documentation
+    #     #original
+    #     #nn.init.xavier_uniform_(self.qkv_proj.weight)
+
+    #     #Added 
+    #     # for l in self.k_projections:
+    #     #     nn.init.xavier_uniform_(l)
+    #     # for l in self.v_projections:
+    #     #     nn.init.xavier_uniform_(l)
+    #     # nn.init.xavier_uniform_(self.o_proj.weight)
+
+    #     #self.qkv_proj.bias.data.fill_(0)
+    #     #self.o_proj.bias.data.fill_(0)
+
+
+
+    def forward(self, x, mask=None, return_attention=False):
```

Comment:
Wygląda ok, można pomyśleć nad deduplikacją względem MultiheadAttention i zmniejszeniem liczby zmian wymiarów (np. użyć view zamiast reshape tam gdzie to możliwe)